### PR TITLE
Introduce and start using microsecond sleep on Linux

### PR DESCRIPTION
**Pull request by ndren**
This is an attempt to help with https://github.com/minetest/minetest/issues/13399#issuecomment-1500993162.

The code previously had something like x/1000*1000 when it came to sleep times which is not particularly useful since the type of x was u64, so this was usually not the same as x.

This code now sleeps for x time on Linux etc., while keeping the same behaviour as before on Windows (which does not provide usleep).

Could someone affected by the FPS issue test whether this improves the issue on platforms that provide usleep?

### CODE REVIEW

**Comment by SmallJoker** [comment]
I have yet not tested the PR but sleeping is not too precise on non-realtime systems unless they're busy waits. Here's an explanation for the implementation on Windows: https://stackoverflow.com/a/11456112

The code looks fine and could IMO be merged like that, even though it would not change much or anything at all.

**Comment by Desour** [comment]
It would also be good to sleep with µs precision on semaphores, see also: <https://github.com/minetest/minetest/pull/13370#discussion_r1166033506>

---

For sleeping, we could use `std::this_thread::sleep_for` and leave the implementation to the cpp std lib implementation.

In the thread that Krock linked, someone mentions `SetWaitableTimer`, which allows to specify values with 100ns precision. For semaphores I guess one could wait on both the semaphore and the timer at once, see also: <https://learn.microsoft.com/en-us/windows/win32/sync/using-waitable-timer-objects>, <https://learn.microsoft.com/en-us/windows/win32/sync/waiting-for-multiple-objects>
(Idk anything about this stuff, but maybe these links help someone.)

**Comment by sfan5** [comment]
> This is an attempt to help with [#13399 (comment)](https://github.com/minetest/minetest/issues/13399#issuecomment-1500993162).

While these changes do not hurt it'd be good to have test results so we can _reliably_ tell that it actually helps.

File: src/porting.h [hunk 1]
```diff
@@ -54,6 +54,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 	#include <windows.h>
 
 	#define sleep_ms(x) Sleep(x)
+	#define sleep_us(x) Sleep(x/1000)
```
Comment (sfan5) [inline src/porting.h:57]: should be brackets around `x` to not get bitten by evaluation order

**Comment by ndren** [comment]
I agree test results would be useful. I don't notice the FPS jitter on my hardware. If someone could measure and check whether this change does anything to the overall precision, that would be great.

**Comment by ndren** [comment]
One thing I did test was the precision of usleep, and I have found it is consistently measurably lower than a millisecond by at least 5x, even without e.g. linux-rt. (Arch Linux, linux-lts 6.1.25-1-lts)

**Comment by ndren** [comment]
> > This is an attempt to help with [#13399 (comment)](https://github.com/minetest/minetest/issues/13399#issuecomment-1500993162).
> 
> While these changes do not hurt it'd be good to have test results so we can _reliably_ tell that it actually helps.

@sfan5 I have tested the changes with VSync (60 FPS monitor).
Before PR: FPS 60-61 (variable, 60 FPS about two thirds of the time), drawtime 3ms-10ms, dtime jitter 3% with spikes of 30%.
After PR: FPS 59 (constant), drawtime 1ms-4ms, dtime jitter 0.3% with spikes of 15%.

It also lowered peak CPU usage from 25% to 18%, as well as getting better results for normal CPU usage though I have difficulty measuring that accurately.

I think this is a clear improvement. Can this please be considered for merging at some point? Thanks.

**Pull request merged**